content/en/blog/_posts/2022-04-19-non-graceful-node-shutdown.md

```diff
@@ -0,0 +1,96 @@
+the control plane makes a replacement Pod for that StatefulSet on a different running node.
+If the original node has failed and does not come up, those stateful pods would be stuck in a
+terminating status on that failed node indefinitely.
+
+```
+$ kubectl get pod -o wide
+NAME    READY   STATUS        RESTARTS   AGE   IP           NODE                      NOMINATED NODE   READINESS GATES
+web-0   1/1     Running       0          100m   10.244.2.4   k8s-node-876-1639279816   <none>           <none>
+web-1   1/1     Terminating   0          100m   10.244.1.3   k8s-node-433-1639279804   <none>           <none>
+```
+
+## Try out the new non-graceful shutdown handling
+
+To use the non-graceful node shutdown handling, you must enable the `NodeOutOfServiceVolumeDetach`
+[feature gate](/docs/reference/command-line-tools-reference/feature-gates/) for the `kube-controller-manager`
+component.
+
+In the case of a node shutdown, you can manually taint that node as out of service. You should make certain that
+the node is truly shutdown (not in the middle of restarting) before you add that taint. You could add that
+taint following a shutdown that the kubelet did not detect and handle in advance; another case where you
+can use that taint is when the node is in a non-recoverable state due to a hardware failure or a broken OS.
+The values you set for that taint can be `node.kubernetes.io/out-of-service=nodeshutdown: "NoExecute"`
+or `node.kubernetes.io/out-of-service=nodeshutdown:" NoSchedule"`.
+Provided you have enabled the feature gate as I mentioned earlier, setting the out-of-service taint on a Node
+means that pods on the node will be deleted unless if there are matching tolerations on the pods.
+Persistent volumes attached to the shutdown node will be detached, and for StatefulSets, replacement pods will
+be created successfully on a different running node.
+
+```
+$ kubectl taint nodes <node-name> node.kubernetes.io/out-of-service=nodeshutdown:NoExecute
+
+$ kubectl get pod -o wide
+NAME    READY   STATUS    RESTARTS   AGE    IP           NODE                      NOMINATED NODE   READINESS GATES
+web-0   1/1     Running   0          150m   10.244.2.4   k8s-node-876-1639279816   <none>           <none>
+web-1   1/1     Running   0          10m    10.244.1.7   k8s-node-433-1639279804   <none>           <none>
+```
+
+Note: Before applying the out-of-service taint, you **must** verify that a node is already in shutdown or power off state (not in the middle of restarting), either because the user intentionally shut it down or the node is down due to hardware failures, OS issues, etc.
+
+Once all the workload Pods that are linked to the out-of-service node are moved to a new running node, and the shutdown node has been recovered, you should remove
+that taint on the affected node. If you know that the node will not return to service, you could instead
```

Comment:
on the affected node after the node is recovered

Reply:
Done